Implementing PCA in Python with Scikit-Learn
------------------------------------------------------------------

With the availability of high performance CPUs and GPUs, it is pretty much possible to solve every regression, classification, clustering and other related problems using machine learning and deep learning models. However, there are still various factors that cause performance bottlenecks while developing such models. 

Large number of features in the dataset is one of the factors that affect both the training time as well as accuracy of machine learning models. You have different options to deal with huge number of features in a dataset.

1. Try to train the models on original number of features, which take days or weeks if the number of features is too high.

2. Reduce the number of variables by merging correlated variables.

3. Extract the most important features from the dataset that are responsible for maximum variance in the output. Different statistical techniques are used for this purpose e.g. linear discriminant analysis, factor analysis, and principal component analysis (PCA).

About Principal Component Analysis
-----------------------------------------------------
Principal component analysis, or PCA, is a statistical technique to convert high dimensional data to low dimensional data by selecting the most important features that capture maximum information about the dataset. The features are selected on the basis of variance that they cause in the output. The feature that causes highest variance is the first principal component. 

The feature that is responsible for second highest variance is considered the second principal component, and so on. It is important to mention that principal components do not have any correlation with each other.

Advantages of PCA
----------------------------
There are two main advantages of dimensionality reduction with PCA.

1> The training time of the algorithms reduces significantly with less number of features.

2> It is not always possible to analyze data in high dimensions. For instance if there are 100 features in a dataset. Total number of scatter plots required to visualize the data would be (100(100-1))/2 = 4950. Practically it is not possible to analyze data this way.

Normalization of Features
-------------------------------------

It is imperative to mention that a feature set must be normalized before applying PCA. For instance if a feature set has data expressed in units of Kilograms, Light years, or Millions, the variance scale is huge in the training set. If PCA is applied on such a feature set, the resultant loadings for features with high variance will also be large. Hence, principal components will be biased towards features with high variance, leading to false results.

Finally, the last point to remember before we start coding is that PCA is a statistical technique and can only be applied to **numeric data. Therefore, categorical features are required to be converted into numerical features before PCA can be applied.

Important Note
>>IN PCA, if we have 10 features, it works on these 10 features and makes 10 new features which are weighted features of each of the above 10 features. Note, each feature is the weighted average of all 10 old features. Also note that the weighted coefficients for all new features is different.

Implementing PCA with Scikit-Learn
---------------------------------------------------

In [3]:
# Importing Libraries
import seaborn as sns  # seaborn library is for data visualisation
import numpy as np
import pandas as pd  
import matplotlib.pyplot as plt
from sklearn import svm, datasets

# import some data to play with
# type your code here
irisdata=sns.load_dataset('iris')

irisdata.sample(10)
# have a look at the attributres(=> X) and Labels(=> y) 

#setosa virginika versicolor

sepal_length  sepal_width  petal_length  petal_width     species
115           6.4          3.2           5.3          2.3   virginica
141           6.9          3.1           5.1          2.3   virginica
77            6.7          3.0           5.0          1.7  versicolor
12            4.8          3.0           1.4          0.1      setosa
47            4.6          3.2           1.4          0.2      setosa
109           7.2          3.6           6.1          2.5   virginica
70            5.9          3.2           4.8          1.8  versicolor
66            5.6          3.0           4.5          1.5  versicolor
142           5.8          2.7           5.1          1.9   virginica
9             4.9          3.1           1.5          0.1      setosa

In [15]:
# Preprocessing data. drop Species to make X. Only extract species to make Y 
# type your code here
X=irisdata.drop('species',axis=1)
y=irisdata['species']


# Train Test Split -> use train_test_split()
from sklearn.model_selection import train_test_split
# type your code here
from sklearn.model_selection import train_test_split
X_train, X_test,y_train,y_test=train_test_split(X,y,test_size=0.20,random_state=0)



#  PCA performs best with a normalized feature set. 
# We will perform standard scalar normalization to normalize our feature set.
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)




# Performing PCA using Scikit-Learn is a two-step process:
# 1. Initialize the PCA class by passing the number of components to 
# the constructor.
# 2. Call the fit_transform() over Training_Data  
#    and only transform() over test data 
# The transform method returns the specified 
# number of principal components.

# Applying PCA
from sklearn.decomposition import PCA
pca = PCA()  
X_train = pca.fit_transform(X_train)  
X_test = pca.transform(X_test)  


In the code above, we create a PCA object named pca. We did not specify the number of components in the constructor. Hence, all four of the features in the feature set will be returned for both the training and test sets.

The PCA class contains explained_variance_ratio_ which returns the variance caused by each of the principal components.

In [16]:
explained_variance=pca.explained_variance_ratio_
explained_variance

array([0.72229951, 0.2397406 , 0.03335483, 0.00460506])

~this analysis could defer for each participant depending on the data :
It can be seen that first principal component is responsible for 72.22% variance. Similarly, the second principal component causes 23.9% variance in the dataset. Collectively we can say that (72.22 + 23.9) 96.21% percent of the classification information contained in the feature set is captured by the first two principal components.

In [17]:
# Let's first try to use 1 principal component to train our algorithm. 
# To do so, execute the following code:

from sklearn.decomposition import PCA
pca=PCA(n_components=1)
X_train=pca.fit_transform(X_train)
X_test=pca.transform(X_test)


In [18]:
explained_variance = pca.explained_variance_ratio_ 
explained_variance #100% achieved

array([0.72229951])

In [19]:
# Training and Making Predictions
# In this case we'll use random forest classification 
# for making the predictions.

from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)  
classifier.fit(X_train,y_train)
#Please note: n_estimators is not specified in RandomforestClassifier
#hence default value of 10 is taken. For this you may get FutureWarning
#TO avoid the warning, either specify n_estimators or suppress warnings


# Predicting the Test set results
y_pred = classifier.predict(X_test)  


# Performance Evaluation
from sklearn.metrics import confusion_matrix  
from sklearn.metrics import accuracy_score

# type your code here
cm=confusion_matrix(y_test,y_pred)
print(cm)  
print('Accuracy ' ,  accuracy_score(y_test, y_pred))

[[11  0  0]
 [ 0 12  1]
 [ 0  1  5]]
Accuracy  0.9333333333333333


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


It can be seen from the output that with only one feature, the random forest algorithm is able to correctly predict 28 out of 30 instances, resulting in 93.33% accuracy. **(This accuracy value will change at train-test dataset changes)

In [24]:
# Results with 2 and 3 Principal Components
#------------------------------------------
# -- self try -- from start of this notebook.

Note : 
1.  With two principal components the classification accuracy decreases to 93.33% compared to 83.33% for 1 component.

2. With three principal components the classification accuracy again increases to 86.66%

From the above experimentation we achieved optimal level of accuracy while significantly reducing the number of features in the dataset. We saw that accuracy achieved with only 1 principal component is equal to the accuracy achieved with three feature set i.e. 96.66%. It is also pertinent to mention that the accuracy of a classifier doesn't necessarily improve with increased number of principal components. From the results we can see that the accuracy achieved with one principal component (66.66%) was greater than the one achieved with two principal components (93.33%).

The number of principal components to retain in a feature set depends on several conditions such as storage capacity, training time, performance, etc. In some dataset all the features are contributing equally to the overall variance, therefore all the principal components are crucial to the predictions and none can be ignored. 

**A general rule of thumb is to take number of principal components that contribute to significant variance and ignore those with lower variance returns.